rfd/0095-application-access-okta-integration.md

```diff
@@ -0,0 +1,315 @@
+---
+authors: Michael Wilson (mike@goteleport.com)
+state: draft
+---
+
+# RFD 95 - Application Access Okta Integration
+
+### Required Approvers
+
+* Engineering @r0mant
+* Security @reed
+* Product: (@xinding33 || @klizhentas)
+
+## What
+
+Allow teleport users to request access to specific applications and groups and access
+applications in Okta.
+
+## Why
+
+Today, Teleport supports [single sign-on with Okta](https://goteleport.com/docs/access-controls/sso/okta/).
+Okta is a popular IdP with our customers, so improving the integration between Okta and
+Teleport will be useful.
+
+## Details
+
+### UX
+
+#### Application Access UI
+
+* After authenticating into Teleport and accessing the "Applications" item in the menu,
+  a list of apps sourced from Okta will appear.
+* When users click on an Okta sourced app in the UI, the application will open in a separate tab.
+* These applications will *not* be behind Teleport's proxy, and will redirect to the proper Okta
+  sourced URL for the application. Users will be taken through Okta's authentication process for
+  this process. If users are already logged into Okta, this will be a transparent passthrough to
+  the application.
+* If users do not have access to an application, the link will be greyed out and an option will
+  be available in the drop down menu to the right that allows users to request access to the
+  application.
+
+#### Teleport CLI
+
+* Okta applications will show up when doing a `tsh apps ls` from the command line.
+* Logging in will **not** work for Okta apps.
+
+#### Access requests
+
+* When a user requests access to an application or group, the existing approval workflow
+  will be used to add this access to the user.
+* Changes will be reflected in the identity provider backend when this occurs.
+
+### High level architecture
+
+The general architecture can be seen below, which is put into more detail later in this document.
+
+```mermaid
+flowchart LR
+  subgraph Teleport
+    direction BT
+    DB[(Teleport Database)]
+    App[Application Service]
+    Requests[Access Requests Service]
+
+    subgraph OktaSvc[Okta Service]
+      BG[Background Synchronization Service]
+    end
+
+    Auth[Auth Service]
+    
+    OktaSvc<-->DB
+    App<-->DB
+    App<-->Requests
+    Requests<-->DB
+  end
+
+  Okta[Okta API]
+  OktaSvc<-->Okta
+```
+
+### Okta service with generic identity provider interface
+
+An Okta service should be introduced that synchronizes Okta applications, users, and
```

Comment:
As you're describing the new `okta_service`, I would give the yaml config example first ("This is how Teleport administrator will configure Okta service") instead of jumping into implementation details.